pkg/client/options.go

```diff
@@ -0,0 +1,417 @@
+		opts.FieldSelector = sel
+	}
+}
+
+// InNamespace is a functional option that sets the Namespace field of
+// a ListOptions struct.
+func InNamespace(ns string) ListOptionFunc {
+	return func(opts *ListOptions) {
+		opts.Namespace = ns
+	}
+}
+
+// UseListOptions is a functional option that replaces the fields of a
+// ListOptions struct with those of a different ListOptions struct.
+//
+// Example:
+// cl.List(ctx, list, client.UseListOptions(lo.InNamespace(ns).MatchingLabels(labels)))
+func UseListOptions(newOpts *ListOptions) ListOptionFunc {
+	return func(opts *ListOptions) {
+		*opts = *newOpts
+	}
+}
+
+// UpdateOptions contains options for create requests. It's generally a subset
+// of metav1.UpdateOptions.
+type UpdateOptions struct {
+	// When present, indicates that modifications should not be
+	// persisted. An invalid or unrecognized dryRun directive will
+	// result in an error response and no further processing of the
+	// request. Valid values are:
+	// - All: all dry run stages will be processed
+	DryRun []string
+
+	// Raw represents raw UpdateOptions, as passed to the API server.
+	Raw *metav1.UpdateOptions
+}
+
+// AsUpdateOptions returns these options as a metav1.UpdateOptions.
+// This may mutate the Raw field.
+func (o *UpdateOptions) AsUpdateOptions() *metav1.UpdateOptions {
+
+	if o == nil {
+		return &metav1.UpdateOptions{}
+	}
+	if o.Raw == nil {
+		o.Raw = &metav1.UpdateOptions{}
+	}
+
+	o.Raw.DryRun = o.DryRun
+	return o.Raw
+}
+
+// ApplyOptions executes the given UpdateOptionFuncs and returns the mutated
+// UpdateOptions.
+func (o *UpdateOptions) ApplyOptions(optFuncs []UpdateOptionFunc) *UpdateOptions {
+	for _, optFunc := range optFuncs {
+		optFunc(o)
+	}
+	return o
+}
+
+// UpdateOptionFunc is a function that mutates a UpdateOptions struct. It implements
+// the functional options pattern. See
+// https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md.
+type UpdateOptionFunc func(*UpdateOptions)
+
+// UpdateDryRunAll is a functional option that sets the DryRun
+// field of a UpdateOptions struct to metav1.DryRunAll.
+var UpdateDryRunAll UpdateOptionFunc = func(opts *UpdateOptions) {
+	opts.DryRun = []string{metav1.DryRunAll}
+}
+
+// PatchOptions contains options for patch requests.
+type PatchOptions struct {
+	// When present, indicates that modifications should not be
+	// persisted. An invalid or unrecognized dryRun directive will
+	// result in an error response and no further processing of the
+	// request. Valid values are:
+	// - All: all dry run stages will be processed
+	DryRun []string
+
+	// Force is going to "force" Apply requests. It means user will
+	// re-acquire conflicting fields owned by other people. Force
+	// flag must be unset for non-apply patch requests.
+	// +optional
+	Force *bool
+
+	// FieldManager is the name of the user or component submitting
+	// this request.  It must be set with server-side apply.
+	FieldManager string
```

Comment:
All of Create|Update|PatchOptions in upstream has `FieldManager`.
In CR, only `PatchOptions` has explicit `FieldManager` field. Should we make it consistent for all of Create|Update|PatchOptions?